### ISSUE TITLE
[BUG] Processing slows down with 96 threads

### ISSUE THREAD

**Issue by jseppanen** [sail-sg/envpool#218]
## Describe the bug

I have a test case that runs different number of threads on a 96-core server and I get roughly these throughput numbers:

```
# python run_slowdown.py
1 thread: 17849.3 SPS
2 threads: 22744.9 SPS
4 threads: 30579.2 SPS
8 threads: 39536.7 SPS
16 threads: 46167.0 SPS
32 threads: 48889.3 SPS
64 threads: 40827.2 SPS
96 threads: 17439.2 SPS
```

The throughput at 96 threads (with 960 envs) is essentially the same as with just 1 thread (and 10 envs).

## To Reproduce

I have a repro in the [slowdown](https://github.com/jseppanen/envpool/tree/slowdown) branch; see [slowdown_envpool.h](https://github.com/jseppanen/envpool/blob/slowdown/envpool/slowdown/slowdown_envpool.h) and [run_slowdown.py](https://github.com/jseppanen/envpool/blob/slowdown/examples/run_slowdown.py) for reproduction.

## Expected behavior

The throughput at 96 threads (with 960 envs) should ideally be nearly 90x faster than with just 1 thread (and 10 envs).

## Screenshots

When running with 96 threads, only one core is running at 100% and other cores are almost idle:

![image](https://user-images.githubusercontent.com/83203/199737975-ec3b36ee-56ab-4549-9993-204b1901d972.png)

## System info

Describe the characteristic of your environment:
 * version 0.6.6 compiled from source ((505e669)[https://github.com/jseppanen/envpool/commit/505e6692facfdbf0dbfcac338924e11bce51cdc5])
 * Envpool version 0.6.6
 * Python version 3.8.13
 * Numpy version 1.22.4
 * Ubuntu 20.04.4 LTS

```python
import envpool, numpy, sys
print(envpool.__version__, numpy.__version__, sys.version, sys.platform)
0.6.6 1.22.4 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10) 
[GCC 10.3.0] linux
```

## Reason and Possible fixes

The env contains about 1.5 KB of state, split across 45 arrays. I suspect that with 96 threads, the main thread is bottlenecked at copying the states, and all workers are idling because of that.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://envpool.readthedocs.io/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by jseppanen** [comment]
Alternatively, if I change the test runner so that that `num_envs = 100` and only vary the number of threads, it still doesn't scale like I'd hope
```
# python run_slowdown_100envs.py
1 thread: 20513.1 SPS
2 threads: 27526.1 SPS
4 threads: 34238.4 SPS
8 threads: 43036.9 SPS
16 threads: 47985.6 SPS
32 threads: 49630.8 SPS
64 threads: 42750.7 SPS
96 threads: 39177.7 SPS
```

**Comment by Trinkle23897** [comment]
I’d recommend using 92 instead of 96 because I use another 4 threads to manage buffer queue

**Comment by Trinkle23897** [comment]
For 256 core machine, I found the optimal setting is 248. The reason is the same. Also, 96-96 is sync stepping, and 100-96 is asynchrony stepping, which will have a significant impact on the performance described in our paper and doc

**Comment by Trinkle23897** [comment]
How did you build the wheel? make bazel-release instead of make bazel-build 
Release has -O3 but build has none

**Comment by jseppanen** [comment]
I built with `bazel run --config=release //:setup -- bdist_wheel`

**Comment by jseppanen** [comment]
I thought it's always sync because I'm not changing `batch_size`? Also, async shouldn't make a big difference because the repro env is a dummy one, where step is just a busy loop.

This is what I get with fixed 92 envs:
```
# python run_slowdown_92envs.py
1 thread: 19709.2 SPS
2 threads: 26977.0 SPS
4 threads: 29500.4 SPS
8 threads: 42539.1 SPS
16 threads: 46338.8 SPS
32 threads: 35137.1 SPS
64 threads: 42157.4 SPS
92 threads: 36879.0 SPS
96 threads: 39223.6 SPS
```

The best scaling is at 16 threads, but even then it's only 2.3x faster than single thread, and beyond 16 threads it starts to get slower again.

**Comment by jseppanen** [comment]
BTW envpool is a wonderful piece of software, thanks for releasing it!

**Comment by Trinkle23897** [comment]
I guess there's some misconfiguration. I'll take a look soon

**Comment by Trinkle23897** [comment]
can you set batch_size in make_gym? Otherwise there are 960 envs step synchronously, which will be super slow

**Comment by jseppanen** [comment]
I tested setting batch_size but it doesn't change anything (the default value is equal to num_envs)

**Comment by jseppanen** [comment]
Update: my busy loop was being optimized out 😂 The fixed version is (76a9850)[https://github.com/jseppanen/envpool/commit/76a9850fcddc69a7b66665e1ca77835b2e3a3316] which gets similar results:
```
# python run_slowdown.py 
1 thread: 15889.2 SPS
2 threads: 23046.2 SPS
4 threads: 32475.8 SPS
8 threads: 40980.4 SPS
16 threads: 50697.3 SPS
32 threads: 56885.9 SPS
64 threads: 51137.9 SPS
92 threads: 49966.1 SPS
96 threads: 48254.2 SPS
```

The poor scaling remains; best speedup compared to single thread is is 3.5x with 32 threads.

However, if I replace the 1.5 KB of state with just a single array of 100 bytes, the scaling improves:
```
# python run_slowdown.py 
1 thread: 15037.3 SPS
2 threads: 28840.8 SPS
4 threads: 54474.3 SPS
8 threads: 107249.9 SPS
16 threads: 195474.2 SPS
32 threads: 291476.8 SPS
64 threads: 429704.4 SPS
92 threads: 483880.2 SPS
96 threads: 483929.0 SPS
```

Now the scaling looks better: 32x speedup with 96 threads. Could it be that the state copying becomes a bottleneck when there is 1.5 KB of state and many envs/threads?

**Comment by Trinkle23897** [comment]
32x is normal. I think the total bytes of state don't matter (Atari uses 4x84x84=28224 uint8 as a **single** state), but the number of state keys matters. They are handled by naive for-loop and we don't have a better way to do this.
